### PR TITLE
ncm-pam: support expressions in the control parameter of the pam rules.

### PR DESCRIPTION
**Pull request by ajf8**
eg. [success=1 default=ignore]

allow options to be passed to pam_add as a list. nlist keys don't allow special characters, or spaces (to preserve ordering of the options).

### CODE REVIEW

File: ncm-pam/src/main/pan/components/pam/config.pan [hunk 1]
```diff
@@ -54,11 +54,16 @@ function pam_add = {
 		ret[service][pamtype] = list();
 	};
 	tail = length(ret[service][pamtype]);
-        options = nlist();
-        if (exists(ARGV[4])) {
-                options = ARGV[4];
+    options = nlist();
```
Comment (stdweird) [inline ncm-pam/src/main/pan/components/pam/config.pan]: fix indentation

File: ncm-pam/src/main/pan/components/pam/config.pan [hunk 2]
```diff
@@ -54,11 +54,16 @@ function pam_add = {
...
+    options_list = null;
+    if (exists(ARGV[4])) {
```
Comment (stdweird) [inline ncm-pam/src/main/pan/components/pam/config.pan]: can you add some annotation to the function describing the existing an dthe new argument

File: ncm-pam/src/main/pan/components/pam/schema.pan [hunk 3]
```diff
@@ -38,9 +38,10 @@ type component_listfile_acl = {
 };
 
 type component_pam_module_stack = {
-	"control" : string with match(SELF, "requisite|required|optional|sufficient")
+	"control" : string with match(SELF, 'requisite|required|optional|sufficient|^\[(.*=.* ?)+\]$')
```
Comment (stdweird) [inline ncm-pam/src/main/pan/components/pam/schema.pan]: fix indentation

Comment (stdweird) [inline ncm-pam/src/main/pan/components/pam/schema.pan]: why is the match not `^(requisite|required|optional|sufficient|\[(.*=.* ?)+\])$`
also `.* ?` makes no sense, you probably meant `[^=]+=[^=]+(\s+[^=]+=[^=]+)*`

File: ncm-pam/src/main/perl/pam.pm [hunk 4]
```diff
@@ -71,8 +71,8 @@ sub Configure {
                   }
               }
 
+              my @o = ();
               if (exists $spec->{options}) {
```
Comment (stdweird) [inline ncm-pam/src/main/perl/pam.pm]: remove the exists

File: ncm-pam/src/main/perl/pam.pm [hunk 5]
```diff
@@ -80,8 +80,13 @@ sub Configure {
                           push(@o, "$kv=$spec->{options}->{$kv}");
                       }
                   }
-                  $opts = join(" ", @o);
               }
+              if (exists $spec->{options_list}) {
```
Comment (stdweird) [inline ncm-pam/src/main/perl/pam.pm]: replace with 
```perl
push(@o, @{$spec->{options_list}}) if $spec->{options_list};
```

File: ncm-pam/src/main/perl/pam.pm [hunk 6]
```diff
@@ -71,8 +71,8 @@ sub Configure {
                   }
               }
 
+              my @o = ();
```
Comment (stdweird) [inline ncm-pam/src/main/perl/pam.pm]: use proper variable name

**Comment by ajf8** [comment]
Thanks for the review @stdweird, hopefully fixed now.

File: ncm-pam/src/main/pan/components/pam/config.pan [hunk 7]
```diff
@@ -18,8 +18,7 @@ include {'components/pam/schema'};
 "/software/components/pam/directory" ?= "/etc/pam.d";
 "/software/components/pam/acldir"    ?= "/etc/pam.acls";
 
-# Add rpm to profile
-include { 'components/${project.artifactId}/config-rpm' };
+"/software/packages" = pkg_repl("ncm-${project.artifactId}", "${no-snapshot-version}-${rpm.release}", "noarch");
 
 # standard functions
 include {'pan/functions'};
```
Comment (stdweird) [inline ncm-pam/src/main/pan/components/pam/config.pan]: no `{}`

File: ncm-pam/src/main/pan/components/pam/config.pan [hunk 8]
```diff
@@ -30,6 +29,7 @@ include {'pan/functions'};
 
 #
 # takes (service, type, control, module, options?)
```
Comment (stdweird) [inline ncm-pam/src/main/pan/components/pam/config.pan]: can you wrap this in an annotation
```pan
@documentation{
  ...
}
```

**Comment by stdweird** [comment]
@ajf8 minor remark left

**Comment by ajf8** [comment]
Minor remarks hopefully fixed too now :+1: thanks

**Comment by stdweird** [comment]
LGTM
@ajf8 thanks!

**Pull request merged**